### PR TITLE
Pdf and tables parsing PoC script - Using Unstructured and Pandas

### PR DESCRIPTION
**Pull request by adivik2000**
This script showcases extracting text from the pdf and if it finds a table, we parse the table as html using pandas and convert each row as page content. Both, table content and regular text retain the metadata(like page no, category etc). Let me know if any changes are required and if so, how must they work?

### CODE REVIEW

**Comment by NirantK** [comment]
Awesome. How to integrate: 

1. Create a `parsers.py` and move the unstructured code there. 
2. Add `unstructured` to `pyproject.toml` — you might have to setup `poetry` locally and then run `poetry add`

*Make a recommendation* on whether we should integrate this with the `chat_complete` or `Conversation`? What are the trade offs? What will the end user workflow/code look like? 

For the recommendation — would prefer pseudo-code instead of English, but can live with English if that works for you :)

**Comment by adivik2000** [comment]
> Awesome. How to integrate:
> 
> 1. Create a `parsers.py` and move the unstructured code there.
> 2. Add `unstructured` to `pyproject.toml` — you might have to setup `poetry` locally and then run `poetry add`
> 
> _Make a recommendation_ on whether we should integrate this with the `chat_complete` or `Conversation`? What are the trade offs? What will the end user workflow/code look like?
> 
> For the recommendation — would prefer pseudo-code instead of English, but can live with English if that works for you :)

```
from agentai.api import chat_complete
from agentai.parsers import Parser
from pydantic import BaseModel, Field

def doc_loader(filename):
     docs = Parser.parse_pdf(filename)
     # Using chroma db as an example
     collection = chroma_client.create_collection(name="my_collection")
     collection.add(
         embeddings=[[1.2, 2.3, 4.5], [6.7, 8.2, 9.2]],
         documents=["This is a document", "This is another document"],
         metadatas=[{"source": "my_source"}, {"source": "my_source"}],
         ids=["id1", "id2"]
     )

class Query(BaseModel):
    query_embedding: List[int] = Field(..., description="Embedding for the query to srach")
    query_text: Optional[str] = Field(..., description="Simplified query from the user to search")
    k: str = Field(..., description="The number of results requested")

@tool(registry=search_docs_registry)
def get_best_documents(query: Query) -> List[Document]:

    doc_search = results = collection.query(
         query_embeddings=[query.query_embedding]
         query_texts=[query.query_text],
         n_results=query.k
     )
    return doc_search

# From SQLite Document Code    
agentai_functions = [json.loads(func.json_info) for func in [get_best_documents]]

agent_system_message = """You are ChinookGPT, a helpful assistant who gets answers to user questions from the Chinook Vector Database.
Be as precise as possible to your users
Begin!"""

doc_conversation = Conversation()
doc_conversation.add_message(role="system", content=agent_system_message)
doc_conversation.add_message("user", "Hi, what are the top 3 features of layoutparser?")
assistant_message = chat_complete(
    conversation=doc_conversation, functions=agentai_functions, model=GPT_MODEL, callable_function=get_best_documents
)

doc_conversation.display_conversation(detailed=True)
```

What do you think? This way, an extra embedding call can be saved.

File: pyproject.toml [hunk 1]
```diff
@@ -21,6 +21,7 @@ termcolor = "^2.3.0"
 typing-extensions = "^4.6.3"
 loguru = "^0.7.0"
 pydantic = "^1.10.9"
+unstructured = {version="^0.9.0", extras=["pdf"]}
```
Comment (NirantK) [inline pyproject.toml:24]: Add `Chroma` to `pyproject.toml` please!

File: agentai/vectordb.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import uuid
+from pydantic import BaseModel, Field
+from typing import List, Optional
+from agentai.annotations import tool, ToolRegistry
+
+
+class Query(BaseModel):
+    query_embedding: List[int] = Field(..., description="Embedding for the query to srach")
+    query_text: Optional[str] = Field(..., description="Simplified query from the user to search")
+    k: str = Field(..., description="The number of results requested")
```
Comment (NirantK) [inline agentai/vectordb.py:12]:
```suggestion
class Query(BaseModel):
    query_embedding: Optional[List[int]] = Field(..., description="Embedding for the query to search")
    query_text: Optional[str] = Field(..., description="Simplified query from the user to search")
    k: int = Field(..., description="The number of results requested")
```

Query Embedding can possible be created by the library and not be part of the user query. Why is a number `str`?

File: agentai/testing.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+from api import chat_complete
```
Comment (NirantK) [inline agentai/testing.py:1]: Rewrite this using `pytest` please? 

Add `pytest` as a dev dependency to pyproject (refer Poetry docs!)

File: agentai/conversation.py [hunk 4]
```diff
@@ -51,7 +51,7 @@ def get_history(self) -> List[Message]:
 
         total_tokens = 0
         # Iterate 2 at a time to avoid cutting in between a (prompt, response) pair
-        for i in range(len(self.history) -1, -1, -2):
+        for i in range(len(self.history) - 1, -1, -2):
```
Comment (NirantK) [inline agentai/conversation.py:54]: Is this a typo? Why add a space here?

**Review by NirantK — CHANGES_REQUESTED** [review]
I'll structure this into smaller issues today and tomorrow. Please start with the test changes. Each in a small PR: 
1. Move them to pytest 
2. Add to poetry dev dependency
3. Add to Github action

File: agentai/conversation.py (hunk 4 above)
Comment (adivik2000) [reply agentai/conversation.py:54]: This was done by black formatter. Will check

File: agentai/vectordb.py (hunk 2 above)
Comment (adivik2000) [reply agentai/vectordb.py:12]: Sorry sorry. Took the pseudo code I wrote and forgot to change this.

File: pyproject.toml (hunk 1 above)
Comment (adivik2000) [reply pyproject.toml:24]: this is done

**Review by NirantK — APPROVED** [review]

**Pull request merged**